### PR TITLE
GroupByProc: remove special-case timebin support

### LINKED ISSUE #828: groupby: remove special-cased time bin handling
Once #535 is complete, we can get rid of the special-case time bin handling in the groupby proc by adding zql sugar that turns `-every X` into a groupby key `ts=Time.trunc(ts, X)`. 

This includes generalizing the "early timebin output" feature that is present in every-groupby, which outputs time-bins as soon as input record time has advanced past a bin. This early-emit will be controlled by a new `-sorted` flag that will tell groupby that the input is sorted on the primary groupby key, and that it is therefore safe to do early-emit.

### PR DESCRIPTION
**Pull request by henridf**
re-spin of #869, this time with fancy AST analysis implemented by @nwt 
closes #828

### CODE REVIEW

File: zqd/search/search.go [hunk 1]
```diff
@@ -177,6 +177,5 @@ func launch(ctx context.Context, query *Query, reader zbuf.Reader, zctx *resolve
 	if span == (nano.Span{}) {
 		span = nano.MaxSpan
 	}
-	reverse := query.Dir < 0
-	return driver.Compile(context.Background(), query.Proc, reader, reverse, span, zap.NewNop())
+	return driver.Compile(context.Background(), query.Proc, reader, "ts", true, span, zap.NewNop())
```
Comment (nwt) [inline zqd/search/search.go:180]:
```suggestion
	// Records in a zqd filestore are sorted by descending ts (in zqd/storage/filestore.(*Storage).write).
	return driver.Compile(context.Background(), query.Proc, reader, "ts", true, span, zap.NewNop())
```

**Comment by nwt** [comment]
Maybe change the PR title to "GroupByProc: remove special-case time-binning support"?

**Comment by nwt** [comment]
Even better: "GroupByProc: replace time-binning with computed grouping key".

**Comment by nwt** [comment]
@henridf: I think the proper fix for the TestMarkdownExamples/zql/docs/aggregate-functions/README.md3 failure is to change the ZQL to `every 5min count() | sort ts` at https://github.com/brimsec/zq/blame/1bd93db184ab103e661b3bcd1b596bea42cfb4b6/zql/docs/aggregate-functions/README.md#L84.

File: proc/groupby.go [hunk 2]
```diff
@@ -131,63 +126,73 @@ type GroupByAggregator struct {
...
+		reducerDefs:  params.reducers,
+		builder:      params.builder,
+		keyRows:      make(map[int]keyRow),
+		table:        make(map[string]*GroupByRow),
+		reverse:      c.Reverse,
+		logger:       c.Logger,
+		limit:        limit,
+		recordSortFn: recordSortFn,
+		valueSortFn:  valueSortFn,
+		inputSortDir: params.inputSortDir,
```
[the remark this answers is not in the document]
Comment (henridf) [reply proc/groupby.go]: failed build due to two limit fields... will fix.

**Comment by henridf** [comment]
I verified once again that Brim master renders the histogram progressively when pointed at this branch, and that `npm test` passes there.

**Pull request merged**